File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        row, col = len(matrix), len(matrix[0])
        i, j = 0, col -1
        while(i <row and j >=0):
            if matrix[i][j] == target:
                return True
            elif matrix[i][j] > target:
                j -= 1 
            else:
                i += 1

        return False
        # m = len(matrix)
        # n = len(matrix[0])
        # for i in range(m):
        #     if target <= matrix[i][n-1]:
        #         low, high = 0, n-1
        #         while(low <= high):
        #             mid = (low + high)//2
        #             if (matrix[i][mid] == target):
        #                 return True
        #             elif(matrix[i][mid] > target):
        #                 high = mid -1
        #             else:
        #                 low = mid + 1
        # return False

        

File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix.java
class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        int m = matrix.length, n = matrix[0].length;
        
        for(int i =0; i<m; i++){
            
            if(target<=matrix[i][n-1]){
                int low = 0, high = n-1;
                
                while(low<=high){
                    int mid = (low+high)/2;
                    if(target == matrix[i][mid]) return true;
                    
                    else if (target>matrix[i][mid]){
                        low = mid+1;
                    }
                    else 
                        high = mid-1;
                }
            }
        }
        return false;
    }
}